File: src/main/scala/tdm/featureextractor/features/PageFeatures.scala
package tdm.featureextractor.features

object PageFeatures {
  @SuppressWarnings(Array("org.wartremover.warts.OptionPartial"))
  def apply(version: String,
            language: Option[String],
            header: Option[SectionFeatures],
            body: Option[SectionFeatures],
            footer: Option[SectionFeatures]): PageFeatures = {
    val allSections = List(header, body, footer)
    val nonEmptySections = allSections.collect { case Some(s) => s }

    val tokenCount = nonEmptySections.foldLeft(0)(_ + _.tokenCount)
    val lineCount = nonEmptySections.foldLeft(0)(_ + _.lineCount)
    val emptyLineCount = nonEmptySections.foldLeft(0)(_ + _.emptyLineCount)
    val sentenceCount =
      if (nonEmptySections.forall(_.sentenceCount.isDefined))
        Some(nonEmptySections.foldLeft(0)(_ + _.sentenceCount.get))
      else None

    new PageFeatures(
      version = version,
      language = language,
      tokenCount = tokenCount,
      lineCount = lineCount,
      emptyLineCount = emptyLineCount,
      sentenceCount = sentenceCount,
      header = header,
      body = body,
      footer = footer
    )
  }

  def unapply(arg: PageFeatures): Option[(String, Option[String], Int, Int, Int, Option[Int], Option[SectionFeatures], Option[SectionFeatures], Option[SectionFeatures])] =
    Some(arg.version, arg.language, arg.tokenCount, arg.lineCount, arg.emptyLineCount, arg.sentenceCount, arg.header, arg.body, arg.footer)
}

/**
  * Object recording aggregate features at the page level
  *
  * @param version        The MD5 hash of the page content
  * @param language       The identified page language (if any)
  * @param tokenCount     The total token count for the page
  * @param lineCount      The total line count for the page
  * @param emptyLineCount The empty line count for the page
  * @param sentenceCount  The sentence count for the page
  * @param header         The page header features
  * @param body           The page body features
  * @param footer         The page footer features
  */
class PageFeatures(val version: String,
                   val language: Option[String],
                   val tokenCount: Int,
                   val lineCount: Int,
                   val emptyLineCount: Int,
                   val sentenceCount: Option[Int],
                   val header: Option[SectionFeatures],
                   val body: Option[SectionFeatures],
                   val footer: Option[SectionFeatures]) extends BasicFeatures

File: src/main/java/org/hathitrust/htrc/featureextractor/java/features/PageFeatures.java
package org.hathitrust.htrc.featureextractor.java.features;

public class PageFeatures implements BasicFeatures {
    protected final String version;
    protected final String language;
    protected final int tokenCount;
    protected final int lineCount;
    protected final int emptyLineCount;
    protected final Integer sentenceCount;
    protected final SectionFeatures header;
    protected final SectionFeatures body;
    protected final SectionFeatures footer;

    /**
     * Object recording aggregate features at the page level
     *
     * @param version        The MD5 hash of the page content
     * @param language       The identified page language (if any)
     * @param tokenCount     The total token count for the page
     * @param lineCount      The total line count for the page
     * @param emptyLineCount The empty line count for the page
     * @param sentenceCount  The sentence count for the page
     * @param header         The page header features
     * @param body           The page body features
     * @param footer         The page footer features
     */
    public PageFeatures(String version, String language, int tokenCount, int lineCount,
                        int emptyLineCount, Integer sentenceCount,
                        SectionFeatures header,
                        SectionFeatures body,
                        SectionFeatures footer) {
        this.version = version;
        this.language = language;
        this.tokenCount = tokenCount;
        this.lineCount = lineCount;
        this.emptyLineCount = emptyLineCount;
        this.sentenceCount = sentenceCount;
        this.header = header;
        this.body = body;
        this.footer = footer;
    }

    public String getVersion() {
        return version;
    }

    public String getLanguage() {
        return language;
    }

    @Override
    public int getTokenCount() {
        return tokenCount;
    }

    @Override
    public int getLineCount() {
        return lineCount;
    }

    @Override
    public int getEmptyLineCount() {
        return emptyLineCount;
    }

    @Override
    public Integer getSentenceCount() {
        return sentenceCount;
    }

    public SectionFeatures getHeader() {
        return header;
    }

    public SectionFeatures getBody() {
        return body;
    }

    public SectionFeatures getFooter() {
        return footer;
    }
}
